gpt_index/optimzation/optimizer.py

```diff
@@ -0,0 +1,102 @@
+"""Optimization related classes and functions."""
+import logging
+from typing import Dict, Optional
+
+from gpt_index.embeddings.base import BaseEmbedding
+from gpt_index.embeddings.openai import OpenAIEmbedding
+from gpt_index.indices.query.embedding_utils import get_top_k_embeddings
+from gpt_index.indices.query.schema import QueryBundle
+from gpt_index.indices.response.builder import TextChunk
+
+
+class Optimizer:
+    """Optimization of a text chunk given the query by shortening the input text."""
+
+    def __init__(
+        self,
+        split_mode: str,
+        comparison_mode: str,
+        embed_model: Optional[BaseEmbedding] = None,
+        cutoffs: Dict[str, float] = {},
+    ):
+        """Optimizer class that is passed into BaseGPTIndexQuery.
+
+        Should be set like this:
+
+        .. code-block:: python
+        from gpt_index.optimization.optimizer import Optimizer
+        optimizer = Optimizer(
+                        split_mode="sentence",
+                        comparison_mode="embedding",
+                        cutoffs = {"percentile": 0.5}
+                        # this means that the top 50% of sentences will be used.
+                        # Alternatively, you can set the cutoff using a threshold
+                        # on the similarity score.
+                        # cutoffs = {"threshold": 0.7}
+                    )
+
+        response = index.query(
+            "<query_str>", optimizer=optimizer
+        )
+        """
+        self._split_mode = split_mode
+        self._comparison_mode = comparison_mode
+        self._embed_model = embed_model or OpenAIEmbedding()
+        self._cutoffs = cutoffs
+
+    def optimize(self, query_bundle: QueryBundle, textChunk: TextChunk) -> TextChunk:
+        """Optimize a text chunk given the query by shortening the input text."""
+        import nltk.data
+
+        if self.is_valid_mode():
+
+            if self._split_mode == "sentence":
+                try:
+                    nltk.data.find("tokenizers/punkt")
+                except LookupError:
+                    nltk.download("punkt")
+                tokenizer = nltk.data.load("tokenizers/punkt/english.pickle")
+                split_text = tokenizer.tokenize(textChunk.text)
+                if self._comparison_mode == "embedding":
+                    query_embedding = self._embed_model.get_agg_embedding_from_queries(
```

Comment:
so, this still calls the embedding API right? does this mean that if the optimizer is supplied, embedding token usage will go up and LLM token usage will go down? (i didn't see that in the screenshot you sent me)

Reply:
Yeah that's the idea. Not sure if the embedding token counter is working correctly in that case? Might need closer investigation.

Reply:
Ah of course it's not summing up the two embed model counts. I will fix that.

Reply:
ah ok yeah let's update the playground exmaple too